android/jni/com/mapswithme/maps/MwmApplication.cpp

```diff
@@ -13,18 +13,18 @@
 extern "C"
 {
   JNIEXPORT void JNICALL
-  Java_com_mapswithme_maps_MwmApplication_nativeInit(
-      JNIEnv * env, jobject thiz,
-      jstring apkPath, jstring storagePath, jstring tmpPath, jstring obbGooglePath,
-      jstring flavorName, jstring buildType, jboolean isYota, jboolean isTablet)
+  Java_com_mapswithme_maps_MwmApplication_nativeInit(JNIEnv * env, jobject thiz,
+                                                     jstring apkPath, jstring settingsPath, jstring tmpPath, jstring obbGooglePath,
+                                                     jstring flavorName, jstring buildType, jboolean isTablet)
   {
-    android::Platform::Instance().Initialize(
-        env, apkPath, storagePath, tmpPath, obbGooglePath, flavorName, buildType, isYota, isTablet);
-
-    LOG(LDEBUG, ("Creating android::Framework instance ..."));
+    android::Platform::Instance().Initialize(env, apkPath, settingsPath, tmpPath, obbGooglePath, flavorName, buildType, isTablet);
+  }
 
-    if (!g_framework)
-      g_framework = new android::Framework();
+  JNIEXPORT void JNICALL
+  Java_com_mapswithme_maps_MwmApplication_nativeStart(JNIEnv * env, jobject thiz, jstring mapsPath)
+  {
+    android::Platform::Instance().SetStoragePath(jni::ToNativeString(env, mapsPath));
+    g_framework = new android::Framework();
```

Comment:
Even more - you probably should call reset(nullptr) first, to avoid situation when new framework is already created but old one is still exist.

> On Oct 7, 2015, at 03:06, ygorshenin <notifications@github.com> wrote:
> 
> In android/jni/com/mapswithme/maps/MwmApplication.cpp <https://github.com/mapsme/omim/pull/131#discussion_r41372283>:
> 
> >  
> > -    if (!g_framework)
> > -      g_framework = new android::Framework();
> > +  JNIEXPORT void JNICALL
> > +  Java_com_mapswithme_maps_MwmApplication_nativeStart(JNIEnv * env, jobject thiz, jstring mapsPath)
> > +  {
> > +    android::Platform::Instance().SetStoragePath(jni::ToNativeString(env, mapsPath));
> > +    g_framework = new android::Framework();
> Смысл unique_ptr в том что если все-таки будет повторная инициализация предыдущий экземпляр полностью завершил все операции и удалился. В противном случае они будут интерферировать приводя к чудным результатам.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/mapsme/omim/pull/131/files#r41372283>.
>